### DISCUSSION TITLE
new ideas

### DISCUSSION THREAD

**Discussion by brusand** [CoderLine/alphaTab#499] [Ideas]
. display mesure, metronome and notes in duration time instead of note (for beginers or non solfegian)
. have a real feeback from mic during playing 

my 2 cents

**Comment by Danielku15** [comment]
I would highly appreciate if you can phrase out full sentences and give an introduction to your messages. Just throwing some buzz-phrases is not really a convenient way of communicating with each other. I am not sure what you expect from such very vague phrases to receive back. 

If you have concrete feature requests it would be good to fully draw the picture of what you would like to have, why and how it could look like. Consider that I am not familiar with the things you would like to have and your goal is to give me all information to grasp the details of your request. e.g. I cannot really grasp in any way how a request "have a real feeback from mic during playing" could relate to alphaTab and what your actual idea is. Why would a microphone be relevant for a component like alphaTab which displays music notation and plays it? You want to record notation with audio detection? Or you want to be able to play via microphone and alphaTab should detect+display the notes played?

**Reply by brusand** [reply]
Hi Daniel,
Sorry too busy to explain my ideas. First congratulation for your great job. I knew Haxe since the first version so many years ago and i am very please to see the new developpement.
I am a guitarist beginer and have a lot of difficuties with rythm and solfegio. So as you have some great components, i image to combien some of them to other tools to improve my efforts. 
1 / display mesure in duration time
in each mesure display each beat time (per ex in a 4/4, divide each measure in 4 columns) et display each note in a row with the  correct duration (per ex: a quarter dotted, a rectangle start on the first beat and stop at the half of the second beat)
i have some friend who played very well but don t read the notes and spends a lot time to retrieve the melody by ear.


2/ custom metronnome
ear beats is not suffisant, to understand the rythm between 2 beats, we could have a second beats count between. Or to have a circle metronome to visualise of top of the score between playing.


3/ strumming pattern
you can enter a pattern profile like down-, down-up-down--down-up. and transtale immediatly in textTab and play the rythm


4/ a tools to have a realtime feeback during playing score.
when i try to play , i want to have a real feedback of what i am playin not exactly the real note or real chord but only rythm if i am on/before or after  the beat. Per exemple having a line to show events on the microphone. One other option will be to have an input like websockets to a external application to recognize the note and send events to alphaTab via websockets.

I understand that alphaTab is not guitarPro and nor an editor, just a tool to render and playing score
sorry for my english i am not fluent.
My 2 cents

**Reply by Danielku15** [reply]
1. This sounds like an interesting idea which might be quite useful for usages in teaching software build with alphaTab. I am not sure if I fully get the same picture like you on how it should look. Do you maybe have some references following this practice? From what I understand the main idea would be to draw some additional guides/annotations which helps readers to relate the note durations to the time signature of the measure. This sounds especially helpful when playing along the metronome as beginner when you cannot grasp the whole bar timings from the standard notation. The things we would have to clarify here is rather how should it actually look that it fits the needs of software trying to train/explain these timings. I am thinking also of the possibility to maybe "ask" alphaTab via an API call for the positions of a given input time/tick list. Then these annotations could be also generated outside of alphaTab as an overlay. 

2. This one I do not understand well. A metronome is meant to tick in the beat interval given by the time signature and tempo. Of course alphaTab could also "double" this timing and make every second tick sound different to get a "1 (and) 2 (and) 3..." feeling. I haven't seen such a feature/practice in any music notation software so far. Building a different visual metronome is something outside the alphaTab responsibility as it is not part of the music sheet itself. https://github.com/CoderLine/alphaTab/issues/450 will cover a feature to "obtain" metronome tick events when they were played via event. This might already allow building some custom metronome extensions on top. 

3. Strumming pattern can be defined via [pick strokes](https://www.alphatab.net/docs/showcase/effects#pick-stroke) and [brush strokes](https://www.alphatab.net/docs/showcase/effects#brush-strokes). In Guitar Pro you can find them here in the palette: 
![image](https://user-images.githubusercontent.com/674916/104109887-8ef59100-52d2-11eb-9a09-6043d097c3ea.png) If you refer to "alphaTex" as input, the pick strokes are defined as [beat effects](https://www.alphatab.net/docs/alphatex/beat-effects#simple-effects) `su` or `sd`. Brushes seem to be missing in alphaTex. I added #505 for that. 

4. This sounds way beyond the scope/responsibility of alphaTab but should be rather implemented in the product using alphaTab. alphaTab could provide some features to easier place annotations or reserve additional space for custom annotations in the rendered music sheet. But detecting audio from a microphone and displaying it somehow whether the person practicing is playing in time or out-of-sync is the job of a guitar training software.

**Reply by brusand** [reply]
thanks Daniel 
1/ and 4/ it is like yousician [https://www.google.com/url?sa=i&url=https%3A%2F%2Fwww.pinterest.com%2Fpin%2F675117800371377504%2F&psig=AOvVaw1J2Ke81oTrufNNSbL2Lqur&ust=1610392715593000&source=images&cd=vfe&ved=0CAIQjRxqFwoTCMiDxMOKku4CFQAAAAAdAAAAABAT](url) . They can catch notes played by the player via the microphone and display if the player match the rythm or not. (it is not the process display the notes i am playing). but i am not fan of the metronome (the reasaon of 2/)

For a beginer it is difficult to separate rythm inside a beat. so i imagine somthing like an analogique watch with 3 hands one for hour beat,  a second for minutes and a third for seconds. I imagine like a clock with 4 hours 16 minutes and 32 seconds.
one option a the metronome will be to have an accent on the half of the 4th beat of a mesure (time to change chords lol) 

for the 3/ the possibility to enter rapidly a rythm pattern without guitar pro , translate dynamicaly with alphaText and play in a loop with alphaTab.

i understand that AlphaTab is more like yousicians

**Comment by brusand** [comment]
hi Daniel,

i start to develop my metronome in Haxe with featherUI, is it possible to add a alphaText or alphaTab to my app ? is it possible to sync alphatab with my metronome ? is possible to script alphatab form outside ?

**Reply by Danielku15** [reply]
alphaTab is a library or SDK if you like. You can learn about everything on the website, the tutorial gives you a starting point on how to use and interact with alphaTab: https://www.alphatab.net/docs/tutorial-web/introduction

Once you understood the basics you can refer to the API reference documentation to learn about all available options, events and API calls to interact with alphaTab: https://www.alphatab.net/docs/reference/api

For further assistance you will have to explain more precisely what you have already, what you would like to actually implement and what alphaTab might be missing or what exactly you want to do with alphaTab.

**Reply by brusand** [reply]
has i am developping in haxe pour multi target (mac os first), my first question is can i integrate aphatab or alphatext via haxelib ?
my first need is to copy a mesure from guitar pro (or alphatab) and paste the selection in my lick area and display the selection.
second need can i control (play/start/pause) the selection from my ui and sync the alphaTab beats with my metronome ?

**Reply by Danielku15** [reply]
> my first question is can i integrate aphatab or alphatext via haxelib ?

No, alphaTab is not distributed via haxelib, it is not a Haxe library. I recommend checking some resources on how to use NPM libs in Haxe. You could try tools like https://github.com/elsassph/hxtsdgen to translate the TypeScript definition to a Haxe definition for better code integration. I recommend asking over at the Haxe Community what is the current approach nowadays to integrate an external JavaScript library which provides a TypeScript definition file. 

> my first need is to copy a mesure from guitar pro (or alphatab) and paste the selection in my lick area and display the selection.

Copy paste from Guitar Pro is not supported in alphaTab, there is no feature for decoding whatever Guitar Pro might copy into the clipboard. Actually it seems this is not even possible. When you copy something in Guitar Pro, there is nothing actually copied to the Clipboard of the Operating System. The copy/paste functionality fully lives within Guitar Pro and there is no way to paste it into an other application. You can test this by using a clipboard viewer application or by simply: 

1. Copying some text  into the clipboard. The text is now in the clipboard. 
2. Go to guitar pro and also copy some notes/beats/measures. 
3. Go to the editor of your choice and try to paste -> Still your text is in the clipboard and will be pasted. 
4. Create a new Guitar Pro file in the still open Guitar Pro instance and paste. -> It will paste the notes/beats/measures you have copied. 

You will have to go the traditional alphaTab approach of saving files in Guitar Pro and then use the provided features to load this file. 

> second need can i control (play/start/pause) the selection from my ui and sync the alphaTab beats with my metronome ?

All this is explained in the Tutorial and on the Website. How to start alphaTab, create UI elements to play/pause/stop and all these things. 
Not sure what "your metronome" is, but alphaTab has a built in metronome. Any code that interacts with your codebase will have to be developed by you.

**Reply by brusand** [reply]
thank you daniel, for copy paste/paste it could be from alpatab, my need is to display/play a score and extract one or 2 mesure to work at different speed (like copy from alpha/TAb and get the mesure via alphaText per example (just need to display 1 or 2 mesure)
but what i understand it will be easy to  dev a web app. thank you daniel i have all my answers
